ARCHITECTURE.md

```diff
@@ -1,113 +1,280 @@
+- Whole assembly stripping has issues with NGO. NGO doesn’t support NetworkBehaviour stripping. Components related to a NetworkObject need to match client and server side, else it’ll mess with your NetworkBehaviour indexing.
+
+After those experimentations, we established new rules for the team:
+
+- Domain based assemblies
+- Use single classes for small components (think the boss room door with a simple on/off state)
+  - If your class never grows too big, having a single `NetworkBehaviour` is perfectly fine and easy to keep the whole class in mind when maintaining it.
+- Use client and server classes (with each pointing to the other) for client/server separation.
+  - Client/Server pair is in the same assembly.
+  - If you start up the game as a client, the server components will disable themselves, leaving you with `{Client}` components executing (make sure you don’t completely destroy the server components, as NGO still requires them for network message sending.
+  - Client would have an m_Server and Server would have an m_Client property.
+  - Server class would own server driven netvars, same for Client with Owner driven netvars.
+  - This way, when reading server code, you do not have to mentally skip client code and vice versa. This helps make bigger classes more readable and maintainable.
+- Use partial classes when the above isn’t possible
+  - Still use the Client/Server prefix for keeping each context in your mind.
+  - Note: you can’t use prefixes for ScriptableObjects that have file name requirements to work.
+- Use Client/Server/Shared separation when you have a 1 to many relationship where your server class needs to send info to many client classes.
+  - This can also be achieved with our NetworkedMessageChannel
+
+You still need to take care of code executing in `Start` and `Awake`: if this code runs contemporaneously with the `NetworkManager`'s initialization, it may not know yet whether the player is a host or client.
+
+### Publisher-Subscriber Messaging
+
+We have implemented a DI-friendly Publisher-Subscriber pattern (see Infrastructure assembly, PubSub folder).
+
+It allows us to send and receive strongly-typed messages in a loosely-coupled manner, where communicating systems only know about the IPublisher/ISubscriber of a given message type. And since publishers and subscribers are classes, we can have more interesting behavior for message transfer, such as Buffered messages (that keep the last message that went through the pipe and gives it to any new subscriber) and networked messaging (see NetworkedMessageChannel section).
```

Comment:
```suggestion
It allows us to send and receive strongly-typed messages in a loosely-coupled manner, where communicating systems only know about the IPublisher/ISubscriber of a given message type. Since publishers and subscribers are classes, we can have more interesting behavior for message transfer, such as Buffered messages (that keep the last message that went through the pipe and gives it to any new subscriber) and networked messaging (see NetworkedMessageChannel section).
```